src/stirling/binaries/stirling_profiler.cc

```diff
@@ -22,111 +22,247 @@
+    // to communicate data to/from eBPF.
+    PX_RETURN_IF_ERROR(source_->Init());
+
+    // Sanity check.
+    if (std::chrono::seconds(FLAGS_time) < source_->SamplingPeriod()) {
+      return error::InvalidArgument(
+          "Profiler will stop before any profiles are collected. Try increasing arg. -time or "
+          "reducing arg. -stirling_profiler_table_update_period_seconds.");
     }
+
+    // We give the source connector a data table to write data into. Note that in the context of
+    // the full "Pixie Edge Module (pem)" (with multiple source connectors) it makes more sense
+    // to have the data tables owned outside of the source connectors.
+    source_->set_data_tables({&data_table_});
+
+    return Status::OK();
   }
 
-  g_data_received = true;
+  Status Stop() {
+    if (source_ != nullptr) {
+      // Cleanup. Important!
+      PX_RETURN_IF_ERROR(source_->Stop());
+    }
 
-  return Status::OK();
-}
+    // More cleanup (invokes dtor of source connector). Prevent Stop() from being invoked twice.
+    source_ = nullptr;
+
+    return Status::OK();
+  }
+
+  Status PrintData() {
+    // This prints a long spew of counts and stack traces.
+    // We will replace this with a pprof proto file writer.
+    // 15x: libc.so;main;foo;bar
+    // 12x: libc.so;main;foo;qux
+    for (const auto& [str, count] : histo_) {
+      LOG(INFO) << count << "x: " << str;
+    }
+    return Status::OK();
+  }
+
+  Status BuildHistogram() {
+    const std::vector<TaggedRecordBatch> tablets = data_table_.ConsumeRecords();
+    const uint64_t num_rows = tablets[0].records[0]->Size();
+
+    auto stack_traces_column = tablets[0].records[kStackTraceStackTraceStrIdx];
+    auto counts_column = tablets[0].records[kStackTraceCountIdx];
+
+    // Build the stack traces histogram.
+    for (uint64_t row_idx = 0; row_idx < num_rows; ++row_idx) {
+      const std::string stack_trace_str = stack_traces_column->Get<types::StringValue>(row_idx);
+      const int64_t count = counts_column->Get<types::Int64Value>(row_idx).val;
+      histo_[stack_trace_str] += count;
+    }
+    return Status::OK();
+  }
+
+  Status StopTransferDataThread() {
+    if (transfer_data_thread_.joinable()) {
+      transfer_enable_ = false;
+      {
+        // Prevent transfer_data_thread_ from invoking TransferData().
+        absl::base_internal::SpinLockHolder lock(&perf_profiler_state_lock_);
+
+        // Join the thread.
+        transfer_data_thread_.join();
+
+        // Transfer any remaining data from eBPF to user space.
+        source_->TransferData(ctx_.get());
+      }
+    }
+
+    // This may be paranoid, but it is possible for StopTransferDataThread() to get
+    // called from the signal handler too. In any case, by the time we are here and from
+    // any entry point, transfer_data_thread_ should have been joined already.
+    if (transfer_data_thread_.joinable()) {
+      return error::Internal("transfer_data_thread_ failed to join.");
+    }
```

Comment:
Can you explain this case? I'm not sure I follow why it's worth this additional check.

Reply:
I'm going to remove this check. TBH my original thought process was to check that the join op (above) had succeeded.

Reply:
Removed that just now.